crates/wasmtime/src/compile/scc.rs

```diff
@@ -0,0 +1,528 @@
+//! Strongly-connected components.
+//!
+//! This module implements [Tarjan's algorithm] for finding strongly-connected
+//! components.
+//!
+//! This algorithm takes `O(V+E)` time and uses `O(V+E)` space.
+//!
+//! Tarjan's algorithm is usually presented as a recursive algorithm, but we do
+//! not trust the input and cannot recurse over it for fear of blowing the
+//! stack. Therefore, this implementation is iterative.
+//!
+//! [Tarjan's algorithm]: https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+
+#![expect(dead_code, reason = "used in upcoming PRs")]
+
+use crate::prelude::*;
+use std::ops::Range;
+use wasmtime_environ::{EntityRef, EntitySet, PrimaryMap, SecondaryMap};
+
+/// A strongly-connected component.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Component(u32);
+wasmtime_environ::entity_impl!(Component);
+
+/// The set of strongly-connected components for a graph of `Node`s.
+pub struct StronglyConnectedComponents<Node>
+where
+    Node: EntityRef,
+{
+    /// A map from a component to the range of `self.component_nodes` that make
+    /// up that component's nodes.
+    components: PrimaryMap<Component, Range<u32>>,
+
+    /// The data storage for the values of `self.components`.
+    component_nodes: Vec<Node>,
+}
+
+impl<Node> StronglyConnectedComponents<Node>
+where
+    Node: EntityRef + std::fmt::Debug,
+{
+    /// Find the strongly-connected for the given graph.
+    pub fn new<I, F, S>(nodes: I, successors: F) -> Self
+    where
+        I: IntoIterator<Item = Node>,
+        F: Fn(Node) -> S,
+        S: Iterator<Item = Node>,
+    {
+        // The resulting components and their nodes.
+        let mut component_nodes = vec![];
+        let mut components = PrimaryMap::<Component, Range<u32>>::new();
+
+        // The DFS index counter.
+        let mut index = NonMaxU32::default();
+
+        // The DFS index and the earliest on-stack node reachable from each
+        // node.
+        let mut indices = SecondaryMap::<Node, Option<NonMaxU32>>::new();
+        let mut lowlinks = SecondaryMap::<Node, Option<NonMaxU32>>::new();
```

Comment:
Just a little optimization thing but perhaps we could use a length hint (perhaps from `Iterator::size_hint`) to pre-allocate using `with_capacity`?